File: app/src/main/java/com/carbonplayer/audio/TrackQueue.java
package com.carbonplayer.audio;


import android.support.v4.util.ArrayMap;

import com.carbonplayer.model.entity.MusicTrack;
import com.carbonplayer.model.entity.ParcelableMusicTrack;

import java.util.ArrayList;
import java.util.List;

/**
 * Singleton track queue which mirrors data in {@link MusicPlayerService}
 */
public class TrackQueue {
    private static TrackQueue instance;
    private ArrayList<MusicTrack> queue;
    private int position;

    public static TrackQueue instance(){
        if(instance==null) instance = new TrackQueue();
        return instance;
    }

    public static TrackQueue instance(List<MusicTrack> tracks, boolean replace){
        if (instance==null)
            instance = new TrackQueue(tracks);
        else if (replace) instance.replace(tracks);

        return instance;
    }

    private TrackQueue(){
        queue = new ArrayList<>();
    }

    private TrackQueue(List<MusicTrack> tracks){
        replace(tracks);
    }

    public void replace(List<MusicTrack> tracks){
        if(tracks instanceof ArrayList)
            queue = (ArrayList<MusicTrack>)tracks;
        else
            queue = new ArrayList<>();
            for(MusicTrack track : tracks) queue.add(track);
    }

    public ArrayList<ParcelableMusicTrack> getParcelable(){
        ArrayList<ParcelableMusicTrack> parcelables = new ArrayList<>();
        int i = 0;
        for(MusicTrack t : queue){
            parcelables.add(new ParcelableMusicTrack(t));
            i++;
        }
        return parcelables;
    }

    public ArrayList<MusicTrack> queue() {return queue;}

    public void insertAtEnd(MusicTrack t){
        queue.add(t);
    }

    public void insertNext(MusicTrack t){
        queue.add(position+1, t);
    }

    public void reorder(int pos, int pnew){
        MusicTrack t = queue.get(pos);
        queue.remove(pos);
        queue.add(pnew, t);
    }

    public void prevtrack(){
        position--;
    }

    public void nexttrack(){
        position++;
    }

    public MusicTrack currentTrack(){
        return queue.get(position);
    }

    public int position(){
        return position;
    }
}

File: app/src/main/java/com/carbonplayer/audio/TrackQueue.kt
package com.carbonplayer.audio


import com.carbonplayer.model.entity.ParcelableTrack
import com.carbonplayer.model.entity.base.ITrack
import com.carbonplayer.utils.parcelable
import io.realm.Realm

/**
 * Track queue which mirrors data in [MusicPlayerService]
 */
class TrackQueue(val callback: TrackQueueCallback) {

    private var queue = mutableListOf<ITrack>()
    var position: Int = 0

    fun replace(tracks: List<ITrack>, pos: Int, noCallback: Boolean = false, local: Boolean = true) {
        queue.clear()
        queue.addAll(tracks)
        position = pos
        if (!noCallback) callback.replace(parcelable(local), pos)
    }

    fun parcelable(local: Boolean = true) = mutableListOf<ParcelableTrack>().apply {
        wrapRealm(local) { realm ->
            queue.forEach {
                this.add(it.parcelable(realm))
            }
        }
    }

    val size: Int get() = queue.size

    fun insertAtEnd(tracks: List<ITrack>, noCallback: Boolean = false, local: Boolean = true) {
        queue.addAll(tracks)
        wrapRealm(local) { realm ->
            if (!noCallback) callback.insertAtEnd(tracks.parcelable(realm))
        }
    }

    fun insertNext(tracks: List<ITrack>, noCallback: Boolean = false, local: Boolean = true) {
        queue.addAll(position, tracks)
        wrapRealm(local) { realm ->
            if (!noCallback) callback.insertNext(tracks.parcelable(realm))
        }
    }


    private inline fun wrapRealm(local: Boolean, crossinline block: (realm: Realm?) -> Unit) {

        if (!local) Realm.getDefaultInstance().executeTransaction({ block(it) })
        else return block(null)
    }

    fun reorder(pos: Int, pnew: Int, noCallback: Boolean = false) {
        val t = queue[pos]
        queue.removeAt(pos)
        queue.add(pnew, t)
        if (!noCallback) callback.reorder(pos, pnew)
    }

    fun remove(pos: Int, noCallback: Boolean = false) {
        queue.removeAt(pos)
        if(!noCallback) callback.remove(pos)
    }


    fun prevtrack() {
        position--
    }

    fun nexttrack() {
        position++
    }

    fun currentTrack(): ITrack {
        return queue[position]
    }

    interface TrackQueueCallback {
        fun replace(tracks: MutableList<ParcelableTrack>, pos: Int)
        fun insertAtEnd(tracks: MutableList<ParcelableTrack>)
        fun insertNext(tracks: MutableList<ParcelableTrack>)
        fun reorder(pos: Int, pnew: Int)
        fun remove(pos: Int)
    }
}
